src/Pos/Wallet/Web/Server/Methods.hs

```diff
@@ -626,69 +626,89 @@ sendMoney sendActions passphrase moneySource dstDistr title desc = do
                (toList entries)
                remains
 
-getHistory
-    :: WalletWebMode m
-    => AccountId -> Maybe Word -> Maybe Word -> m ([CTx], Word)
-getHistory accId skip limit = do
-    accAddrs <- getAccountAddrsOrThrow Ever accId
-    addrs <- forM accAddrs (decodeCIdOrFail . cwamId)
-    cHistory <-
-        do  (minit, cachedTxs) <- transCache <$> getHistoryCache accId
-
-            -- TODO: Fix type param! Global type param.
-            TxHistoryAnswer {..} <- untag @WalletSscType getTxHistory addrs minit
-
-            -- Add allowed portion of result to cache
-            let fullHistory = taHistory <> cachedTxs
-                lenHistory = length taHistory
-                cached = drop (lenHistory - taCachedNum) taHistory
-            unless (null cached) $
-                updateHistoryCache
-                    accId
-                    taLastCachedHash
-                    taCachedUtxo
-                    (cached <> cachedTxs)
-
-            forM fullHistory $ addHistoryTx accId mempty mempty
-    pure (paginate cHistory, fromIntegral $ length cHistory)
+getFullHistory :: WalletWebMode m => ObjectWithHistory -> m ([CTx], Word)
+getFullHistory owh = do
+    accIds <- case owh of
+        OWHWallet cWalId -> getWalletAccountIds cWalId
+        OWHAccount accId -> pure [accId]
+    accAddrs <- concatMapM (getAccountAddrsOrThrow Ever) accIds
+    addrs <- mapM (decodeCIdOrFail . cwamId) accAddrs
+    cHistory <- do
+        (mInit, cachedTxs) <- transCache <$> getHistoryCache owh
```

Comment:
Having a separate independent cache for wallet given that every account in wallet already have its own cache is space-consuming and error-prone. It would be better to leave caches only for account ids, and here, instead of getting cache by `ObjectWithHistory`, call a helper which aggregates cached histories for `accIds` and merges them.